### PR TITLE
Allow specification of max slide for unavailable range

### PR DESCRIPTION
**Pull request by zhen-zen**
https://github.com/acidanthera/bugtracker/issues/979
Some higher slides may be unavailable and sometimes can cause boot failure when generated random slide fall into this range.

### CODE REVIEW

File: Library/OcAfterBootCompatLib/CustomSlide.c [hunk 1]
```diff
@@ -89,11 +89,18 @@ GenerateSlideValue (
...
 
-  do {
-    DivU64x32Remainder (GetPseudoRandomNumber64 (), SlideSupport->ValidSlideCount, &Slide);
-  } while (Slide == 0);
+  //
+  // Handle maximum available slide case.
+  //
+  DivU64x32Remainder (
+    GetPseudoRandomNumber64 (),
+    SlideSupport->ProvideMaxSlide == 0
```
Comment (vit9696) [inline Library/OcAfterBootCompatLib/CustomSlide.c:97]: Ugh, this is still wrong. In this implementation `SlideSupport->ProvideMaxSlide` will be maximum number of slides used. I.e. it says first N slides, but these first slides can be anything, they can be 1, 2, 3, 5, 6 or they can be 1, 2, 3, 5, 9, 100, 101, 102, 103 as `SlideSupport->ValidSlides` is a sparse list of values (!).

What I believe this should really do is to add a check [here](https://github.com/acidanthera/OpenCorePkg/blob/796765c/Library/OcAfterBootCompatLib/CustomSlide.c#L365), replacing it with something like:

```C
if (StartAddr + AvailableSize) != EndAddr
  || (SlideSupport->ProvideMaxSlide > 0 && Slide > SlideSupport->ProvideMaxSlide))
```

Comment (zhen-zen) [reply Library/OcAfterBootCompatLib/CustomSlide.c:97]: Or if we can break the outer for-loop before that line and skip evaluating remaining higher slides?

Comment (vit9696) [reply Library/OcAfterBootCompatLib/CustomSlide.c:97]: Hm, or actually that. Yeah, that's more obvious of course.

File: Library/OcAfterBootCompatLib/CustomSlide.c [hunk 2]
```diff
@@ -89,11 +89,9 @@ GenerateSlideValue (
     return SlideSupport->ValidSlides[0];
   }
 
-  do {
-    DivU64x32Remainder (GetPseudoRandomNumber64 (), SlideSupport->ValidSlideCount, &Slide);
-  } while (Slide == 0);
+  DivU64x32Remainder (GetPseudoRandomNumber64 (), SlideSupport->ValidSlideCount - 1, &Slide);
```
Comment (vit9696) [inline Library/OcAfterBootCompatLib/CustomSlide.c:92]: This change looks wrong to me, as now it avoids the first slide entry in ValidSlides, but it is not necessary 0 (e.g. on Z77 it normally starts with 12, 13, and so on). Please change it back to the original code, that is what we reverse-engineered from Efiboot.

File: Library/OcAfterBootCompatLib/CustomSlide.c [hunk 3]
```diff
@@ -374,6 +380,11 @@ ShouldUseCustomSlideOffset (
     }
   }
 
+  if (SlideSupport->ValidSlideCount < SlideSupport->ProvideMaxSlide) {
```
Comment (vit9696) [inline Library/OcAfterBootCompatLib/CustomSlide.c:383]: I do not know what exactly this does, but it looks like some artifact of the previous change. Please remove.

File: Docs/Configuration.tex [hunk 4]
```diff
@@ -1459,6 +1459,22 @@ \subsection{Quirks Properties}\label{booterpropsquirks}
...
 
+\item
+  \texttt{ProvideMaxSlide}\\
+  \textbf{Type}: \texttt{plist\ integer}\\
+  \textbf{Failsafe}: \texttt{0}\\
+  \textbf{Description}: Provide maximum KASLR slide when higher ones are unavailable.
+
+  This option overrides the maximum slide instead of \texttt{TOTAL\_SLIDE\_NUM} (256)
+  when \texttt{ProvideCustomSlide} is enabled. When generating random slides, this
+  option can avoid bad slides that are unavailable and cause boot failure.
```
Comment (vit9696) [inline Docs/Configuration.tex:1470]: Could you add the following afterwards? It is believed that modern firmwares allocate pool memory from top to bottom, effectively resulting in free memory at the time of slide scanning being later used as temporary memory during kernel loading.

File: Library/OcAfterBootCompatLib/CustomSlide.c (hunk 2 above)
Comment (zhen-zen) [reply Library/OcAfterBootCompatLib/CustomSlide.c:92]: Or if the condition should be `SlideSupport->ValidSlides[Slide] == 0`  ?

Comment (vit9696) [reply Library/OcAfterBootCompatLib/CustomSlide.c:92]: Good question, now I see the issue. Let me recheck EfiBoot.

File: Library/OcAfterBootCompatLib/CustomSlide.c [hunk 5]
```diff
@@ -91,7 +91,7 @@ GenerateSlideValue (
 
   do {
     DivU64x32Remainder (GetPseudoRandomNumber64 (), SlideSupport->ValidSlideCount, &Slide);
-  } while (Slide == 0);
+  } while (Slide == 0); // or SlideSupport->ValidSlides[Slide] == 0
```
Comment (vit9696) [inline Library/OcAfterBootCompatLib/CustomSlide.c:94]: My bad, should be `SlideSupport->ValidSlides[Slide] == 0` indeed.

**Pull request merged**